bin/propolis-server/src/lib/migrate/protocol.rs

```diff
@@ -0,0 +1,323 @@
+    }
+}
+
+impl TryFrom<ProtocolParts> for Protocol {
+    type Error = anyhow::Error;
+
+    fn try_from(value: ProtocolParts) -> Result<Self, Self::Error> {
+        let protocol = match value {
+            ProtocolParts { encoding: Encoding::Ron, version: 0 } => {
+                Self::RonV0
+            }
+            _ => anyhow::bail!(format!(
+                "no protocol matching definition: {:?}",
+                value
+            )),
+        };
+
+        Ok(protocol)
+    }
+}
+
+// Migration offers are of the form "propolis-migrate-encoding/version".
+// Offer strings with multiple versions are comma-delimited.
+
+/// The prefix to strip from a protocol offer to get the encoding and version.
+const PREFIX: &str = "propolis-migrate-";
+
+/// The separator in a protocol offer that separates the encoding from the
+/// version.
+const ENCODING_VERSION_SEPARATOR: char = '/';
+
+/// The delimiter separating offers in a single string.
+const DELIMITER: char = ',';
+
+/// Errors that can arise while parsing a protocol offer string.
+#[derive(Clone, Debug, Error, PartialEq, Serialize, Deserialize)]
+pub enum ProtocolParseError {
+    #[error("protocol string did not begin with propolis-migrate: {0}")]
+    InvalidPrefix(String),
+
+    #[error("protocol string did not have a '/' separator: {0}")]
+    NoEncodingVersionSeparator(String),
+
+    #[error("invalid encoding: {0}")]
+    InvalidEncoding(String),
+
+    #[error("failed to parse protocol version number {0}: {1}")]
+    InvalidVersionNumber(String, String),
+
+    #[error("offered protocol set contained duplicate protocol {0}")]
+    DuplicateProtocolInOffer(String),
+}
+
+/// The set of permissible encodings.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+enum Encoding {
+    /// Encode using Rust Object Notation.
+    Ron,
+}
+
+impl Display for Encoding {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Encoding::Ron => "ron",
+            }
+        )
+    }
+}
+
+impl FromStr for Encoding {
+    type Err = ProtocolParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "ron" => Ok(Encoding::Ron),
+            _ => Err(ProtocolParseError::InvalidEncoding(s.to_owned())),
+        }
+    }
+}
+
+/// A protocol selection.
+//
+// N.B. The ordering of fields in this struct matters! It ensures that the
+//      derived impls of PartialOrd and Ord compare versions before encodings.
+//      This ensures that the negotiation process always selects the latest
+//      version irrespective of whether it has the most preferable encoding.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+struct ProtocolParts {
+    version: u32,
+    encoding: Encoding,
+}
+
+impl ProtocolParts {
+    fn offer_string(&self) -> String {
+        format!(
+            "{PREFIX}{}{ENCODING_VERSION_SEPARATOR}{}",
```

Comment:
I'll be honest, the name binding throws me for a loop, but I'm not sure it'd be any clearer with those passed as args.

Reply:
I think what's weird about it is that it interleaves interpolated references and plain old arguments. Doing one or the other looks less silly. Fixed in 17dbf47.